sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Implementation of a liquidity Pool for Sui.
+///
+/// This solution is rather simple and is based on the example from the Move repo:
+/// https://github.com/move-language/move/blob/main/language/documentation/examples/experimental/coin-swap/sources/CoinSwap.move
+module defi::pool {
+    use sui::Coin::{Self, Coin, TreasuryCap};
+    use sui::Balance::{Self, Balance};
+    use sui::ID::{VersionedID};
+    use sui::SUI::SUI;
+    use sui::Transfer;
+    use sui::TxContext::{Self, TxContext};
+
+    /// The Capability to create new pools.
+    struct PoolCreatorCap has key, store {
+        id: VersionedID
+    }
+
+    /// The Pool token that will be used to mark the pool share
+    /// of a liquidity provider.
+    struct LSP<phantom T> has drop {}
+
+    /// The pool with exchange.
+    struct Pool<phantom T> has key {
+        id: VersionedID,
+        sui: Balance<SUI>,
+        token: Balance<T>,
+        lsp_treasury: TreasuryCap<LSP<T>>
+    }
+
+    /// On init creator of the module gets the capability
+    /// to create new `Pool`s.
+    fun init(ctx: &mut TxContext) {
```

Comment:
Why not just let anyone make a pool? Am I missing something?

Reply:
It is really hard to constrain `Coin<T>`. So if anyone can create a `Pool<T>`, they can issue as many `Coin<LSP<T>>` tokens as they want and that would allow them withdrawing assets from other pools.

There are two ways of addressing this problem:
1. Either adding a Witness type to LSP 
2. Or using Capabilities

I went with the latter because it is simpler.

Reply:
@sblackshear I think I'll go with a witness for this one. Reimplementing Coin and making them 2nd level doesn't sound good to me. After all (and I think it was always known) even with a witness type one can create multiple TreasuryCaps and break the tokenomics, so this is something that we assume people will check.